### PR TITLE
Improve the back pressure deadlock error

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/dotnet/corefx/issues/30894

I'm not a fan of the new message either. Any ideas to phrase it better?

### CODE REVIEW

**Comment by pakrym** [comment]
cc @BrennanConroy

File: src/System.IO.Pipelines/src/Resources/Strings.resx [hunk 1]
```diff
@@ -121,7 +121,7 @@
     <value>Pipe is already advanced past provided cursor.</value>
   </data>
   <data name="BackpressureDeadlock" xml:space="preserve">
-    <value>Advancing examined to the end would cause pipe to deadlock because FlushAsync is waiting.</value>
+    <value>The maximum buffer limit of {0} was exceeded by the writer.</value>
```
Comment (anurse) [inline src/System.IO.Pipelines/src/Resources/Strings.resx]:
```suggestion
    <value>The maximum buffer limit of {0} bytes was exceeded by the writer.</value>
```

File: src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// The exception is thrown when buffer limit is exceeded by the writer.
+    /// </summary>
+    public class BufferLimitExceededException : InvalidOperationException
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs]: As new public API, this should go through the normal API review process.

Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs]: cc @terrajobst

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs]: I edited the original issue too

File: src/System.IO.Pipelines/ref/System.IO.Pipelines.cs [hunk 3]
```diff
@@ -72,6 +76,7 @@ public abstract partial class PipeWriter : System.Buffers.IBufferWriter<byte>
     public readonly partial struct ReadResult
     {
         private readonly object _dummy;
+        private readonly int _dummyPrimitive;
```
Comment (halter73) [inline src/System.IO.Pipelines/ref/System.IO.Pipelines.cs]: What's the _dummyPrimitive for, and why do we need _dummy too?

Comment (pakrym) [reply src/System.IO.Pipelines/ref/System.IO.Pipelines.cs]: Autogenerated. http://blog.paranoidcoding.com/2016/02/15/are-private-members-api-surface.html

Comment (ahsonkhan) [reply src/System.IO.Pipelines/ref/System.IO.Pipelines.cs]: > Autogenerated.

woot :)

File: src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs (hunk 2 above)
Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/BufferLimitExceededException.cs]: SignalR, for example, wanted to tell people to change signalr related limit

File: src/System.IO.Pipelines/src/Resources/Strings.resx [hunk 4]
```diff
@@ -121,7 +121,7 @@
     <value>Pipe is already advanced past provided cursor.</value>
   </data>
   <data name="BackpressureDeadlock" xml:space="preserve">
-    <value>Advancing examined to the end would cause pipe to deadlock because FlushAsync is waiting.</value>
+    <value>The maximum buffer limit of {0} bytes was exceeded by the writer.</value>
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/System.IO.Pipelines/src/Resources/Strings.resx]: > Saying that the limit "was exceeded by the writer" is very misleading.

I agree with this. The error is basically trying to indicate that the pipe can't buffer anymore data. What's harder to convey in an error message is this:

> Didn't consume enough data to push the buffer size under the ResumeWriterThreshold

I'm not sure how you describe that if you never configured the pipe to begin with. You can put it in the error message but it doesn't help you fix the issue.

**Comment by stephentoub** [comment]
@pakrym, what's the status of this?

**Comment by pakrym** [comment]
Waiting on API review for https://github.com/dotnet/corefx/issues/30894

**Comment by stephentoub** [comment]
> Waiting on API review for #30894

Can we close this until that's done? It's otherwise not actually now.

File: src/System.IO.Pipelines/tests/BackpressureTests.cs [hunk 5]
```diff
@@ -38,7 +38,9 @@ public void AdvanceThrowsIfFlushActiveAndNotConsumedPastThreshold()
 
             ReadResult result = _pipe.Reader.ReadAsync().GetAwaiter().GetResult();
             SequencePosition consumed = result.Buffer.GetPosition(31);
-            Assert.Throws<InvalidOperationException>(() => _pipe.Reader.AdvanceTo(consumed, result.Buffer.End));
+
+            var exception = Assert.Throws<InvalidOperationException>(() => _pipe.Reader.AdvanceTo(consumed, result.Buffer.End));
+            Assert.Equal("The maximum buffer limit of 32 bytes was exceeded by the writer.", exception.Message);
```
Comment (stephentoub) [inline src/System.IO.Pipelines/tests/BackpressureTests.cs]: We generally avoid doing this; it'll fail if we ever localize.

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: Fix it when you localize? 😄

Comment (stephentoub) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: Oh the debt we'd have.

File: src/System.IO.Pipelines/src/Resources/Strings.resx (hunk 4 above)
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/Resources/Strings.resx]: nit: {0} byte(s) (in case it's 1 byte - can the maximum ever be 1 byte?)

**Pull request merged**